cider/tests/utils/ArrowArrayBuilder.h

```diff
@@ -0,0 +1,152 @@
+  }
+
+  ArrowArrayBuilder& setTableName(const std::string& table_name) {
+    table_name_ = table_name;
+    schema_->name = table_name.c_str();
+    return *this;
+  }
+
+  ArrowArrayBuilder& setRowNum(int row_num) {
+    if (is_row_num_set_) {  // have set before, throw exception
+      CIDER_THROW(CiderCompileException, "row num have been set!");
+    }
+    is_row_num_set_ = true;
+    row_num_ = row_num;
+    return *this;
+  }
+
+  template <class T>
+  ArrowArrayBuilder& addColumn(const std::string& col_name,
+                               const ::substrait::Type& col_type,
+                               const std::vector<T>& col_data,
+                               const std::vector<bool>& null_data = {}) {
+    if (!is_row_num_set_ ||  // have not set row num, use this col_data's row num
+        row_num_ == 0) {     // previous columns are all empty
+      is_row_num_set_ = true;
+      row_num_ = col_data.size();
+    }
+    ArrowArray* current_array = new ArrowArray();
+    ArrowSchema* current_schema = new ArrowSchema();
+
+    current_schema->name = col_name.c_str();
+    current_schema->format = CiderBatchUtils::convertSubstraitTypeToArrowType(col_type);
+    current_schema->n_children = 0;
+    current_schema->children = nullptr;
+    current_schema->release = CiderBatchUtils::ciderEmptyArrowSchemaReleaser;
+
+    if (col_data.empty()) {
+      // append an empty buffer.
+      array_list_.push_back(nullptr);
+      schema_list_.push_back(current_schema);
+      return *this;
+    } else {
+      // check row num
+      if (row_num_ != col_data.size()) {
+        CIDER_THROW(CiderCompileException, "Row num is not equal to previous columns!");
+      }
+      CHECK_EQ(row_num_, col_data.size());
+      // check null data num
+      if (!null_data.empty()) {
+        CHECK_EQ(row_num_, null_data.size());
+      }
+
+      current_array->length = row_num_;
+      current_array->n_children = 0;
+      current_array->offset = 0;
+      current_array->buffers = (const void**)std::malloc(sizeof(void*) * 2);
+
+      void* null_buf = std::malloc(row_num_ / 8 + 1);
+      std::memset(null_buf, 0xFF, row_num_ / 8 + 1);
+      for(auto i=0;i<null_data.size();i++){
+        if(!null_data[i]){
+          CiderBitUtils::clearBitAt((uint8_t*)null_buf,i);
+        }
+      }
+
+      current_array->buffers[0] = null_buf;
+      current_array->buffers[1] = col_data.data();
+      current_array->n_buffers = 2;
+      current_array->private_data = nullptr;
+      current_array->dictionary = nullptr;
+      current_array->release = CiderBatchUtils::ciderEmptyArrowArrayReleaser;
+
+      array_list_.push_back(current_array);
+      schema_list_.push_back(current_schema);
+    }
+    return *this;
+  }
+
+  std::tuple<ArrowSchema*&, ArrowArray*&> build() {
+    if (!is_row_num_set_) {
+      CIDER_THROW(CiderCompileException, "Invalid build!");
+    }
+    array_->length = row_num_;
+    array_->n_children = array_list_.size();
```

Comment:
`size_t column_num = array_list_.size(); ` use `column_num` would be more clear.